File: 백준/Gold/2636. 치즈/치즈.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

var (
	N, M, cheese int
	arr          [][]int
	dx           = []int{0, 0, 1, -1}
	dy           = []int{-1, 1, 0, 0}
)

func bfs() {
	q := [][]int{{0, 0}}
	visit := make([][]bool, N)
	for i := range visit {
		visit[i] = make([]bool, M)
	}
	visit[0][0] = true
	for len(q) > 0 {
		now := q[0]
		q = q[1:]

		for d := 0; d < 4; d++ {
			nx := now[0] + dx[d]
			ny := now[1] + dy[d]
			if nx >= 0 && nx < N && ny >= 0 && ny < M && !visit[nx][ny] {
				visit[nx][ny] = true
				if arr[nx][ny] == 1 {
					arr[nx][ny] = 0
					cheese--
				} else if arr[nx][ny] == 0 {
					q = append(q, []int{nx, ny})
				}
			}
		}
	}

}

func main() {
	br := bufio.NewReader(os.Stdin)
	bw := bufio.NewWriter(os.Stdout)
	defer bw.Flush()

	fmt.Fscanln(br, &N, &M)
	cheese = 0
	arr = make([][]int, N)
	for i := 0; i < N; i++ {
		arr[i] = make([]int, M)
		for j := 0; j < M; j++ {
			fmt.Fscan(br, &arr[i][j])
			if arr[i][j] == 1 {
				cheese++
			}
		}
	}

	var time, res int
	time = 0

	for cheese > 0 {
		res = cheese
		time++
		bfs()
	}
	fmt.Println(time)
	fmt.Println(res)
}

File: 백준/Gold/2636. 치즈/치즈.java
import java.io.*;
import java.util.*;
public class Main{
    static int N,M,arr[][],cheese, dx[] = {0,0,1,-1}, dy[] = {1,-1,0,0};
    static void bfs(){
        Queue<int[]> q = new LinkedList<>();
        boolean[][] visit = new boolean[N][M];
        q.add(new int[]{0,0});
        visit[0][0] = true;
        while(!q.isEmpty()){
            int now[] = q.poll();

            for(int d=0;d<4;d++){
                int nx = now[0]+dx[d], ny = now[1]+dy[d];
                if(0>nx || nx>=N || 0>ny || ny>=M || visit[nx][ny]) continue;
                visit[nx][ny]=true;
                if(arr[nx][ny]==1){
                    arr[nx][ny] = 0;
                    cheese--;
                    }
                else if(arr[nx][ny]==0) q.add(new int[]{nx,ny});
            }
        }
    }

    public static void main(String[] args)throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder ans = new StringBuilder();
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[N][M];
        cheese =0;

        for(int i=0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            for(int j=0; j<M; j++){
                arr[i][j] = Integer.parseInt(st.nextToken());
                if(arr[i][j]==1){
                    cheese++;
                }
            }
        }

        int time =0, res = 0;
        while(cheese>0){
            res = cheese;
            bfs();
            time++;
        }

        ans.append(time+"\n"+res);
        System.out.println(ans);

    }


}